# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

Define the problem you are finding answers for from the data given.

Business Context
Business Objective
Data Objective
Evaluation

## Executive Summary

If you want to, it's great to use relative links to direct your audience to various sections of a notebook. **HERE'S A DEMONSTRATION WITH THE CURRENT SECTION HEADERS**:

### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [161]:
#Imports:
import numpy as np
import pandas as pd

%matplotlib inline

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [148]:
df_sat_2017 = pd.read_csv('../data/sat_2017.csv')
df_act_2017 = pd.read_csv('../data/act_2017.csv')

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [11]:
df_sat_2017.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [98]:
df_sat_2017.describe()

Evidence-Based Reading and Writing        Math        Total
count                           51.000000   51.000000    51.000000
mean                           569.117647  547.627451  1126.098039
std                             45.666901   84.909119    92.494812
min                            482.000000   52.000000   950.000000
25%                            533.500000  522.000000  1055.500000
50%                            559.000000  548.000000  1107.000000
75%                            613.000000  599.000000  1212.000000
max                            644.000000  651.000000  1295.000000

In [15]:
df_act_2017.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

  Composite  
0      21.0  
1      19.2  
2      19.8  
3      19.7  
4      19.4  
5      22.8  
6      20.8  
7      25.2  
8      24.1  
9      24.2

In [17]:
df_act_2017.describe()

English       Math    Reading    Science
count  52.000000  52.000000  52.000000  52.000000
mean   20.919231  21.173077  22.001923  21.040385
std     2.332132   1.963602   2.048672   3.151113
min    16.300000  18.000000  18.100000   2.300000
25%    19.000000  19.400000  20.475000  19.900000
50%    20.550000  20.900000  21.700000  21.150000
75%    23.300000  23.100000  24.125000  22.525000
max    25.500000  25.300000  26.000000  24.900000

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

No null/missing values from both datasets.

Participation rates need to be converted to float data type for both datasets.

For the SAT dataset, the minimum value for the Math column is 52, while the mean is 547. This seems to indicate an outlier,  further investigation is required.

For the ACT dataset, the Composite column needs to be converted to a float data type.

For the ACT dataset, the minimum value for the Science column is 2.3, while the mean is 21. This seems to indicate an outlier, further investigation is required.

In [96]:
#### 4a. Does the data look complete? 
df_sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null object
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int64(3), object(2)
memory usage: 1.6+ KB


In [97]:
df_act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State            52 non-null object
Participation    52 non-null object
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null object
dtypes: float64(4), object(3)
memory usage: 2.3+ KB


Both datasets appear to be complete. However, we need to investigate outliers in both datasets, validating them against their respective source. Also, we need to determine why the Composite score (from ACT dataset), is not loaded as a float data type.

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

For SAT, both EBRW and Math tests have minimum/maximum possible values of 200/800. For ACT, the 4 subject area and composite scores have a minimum/maximum possible values of 0/36.

For the SAT dataset, this indicates that the outlier Math score of 52 is an error as the minimum score is 200.

Validating the SAT dataset against its source, shows that the Math score for the state of Maryland should be 524 instead of 52.


In [149]:
# Validate SAT CSV data with data from source @ https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/

# Set index (State) and rename columns of SAT CSV and source dataframes to facilitate validation later.

sat_2017_columns = ['Participation', 'EBRW', 'Math', 'Total']

df_sat_2017.set_index('State', inplace=True)
df_sat_2017.columns = sat_2017_columns

In [150]:
# Load source SAT data from tab delimited file into dataframe.
df_sat_2017_source = pd.read_csv('../data/sat_2017_source.txt', sep='\t')
df_sat_2017_source.set_index('State', inplace=True)
df_sat_2017_source.columns = sat_2017_columns
df_sat_2017_source.rename(columns=lambda col_name: col_name + '_src', inplace=True) 

# Join SAT CSV and source dataframes on row index 'State' 
df_sat_2017_combined = df_sat_2017.join(df_sat_2017_source, 'State')


# Compares the CSV and source columns and returns a list of State index values of rows with mismatched columns
def get_mismatched_indexes(df_combined, columns):
    
    mismatched_indexes = []
    
    for column in columns:

        rows = df_combined[df_combined[column] != df_combined[column + '_src']]
        
        if len(rows) > 0:
            mismatched_indexes.extend(rows.index.values)
        
    return list(set(mismatched_indexes))


df_sat_2017_mismatched = df_sat_2017_combined.loc[get_mismatched_indexes(df_sat_2017_combined, sat_2017_columns)]

# Highlight a mismatch between a CSV and source column value
def highlight_mismatched(column):
    
    formats = ['' for i in range(column.size)]
    
    if column.name in sat_2017_columns:
        
        is_mismatched = column != df_sat_2017_mismatched[column.name + '_src']

        return ['background-color: yellow' if mismatched else '' for mismatched in is_mismatched]
    
    return formats

df_sat_2017_mismatched.style.apply(highlight_mismatched)

Before we can validate the ACT CSV dataset against its source, we must ensure that the columns from the CSV dataset and the corresponding column from the source have the same data type.

Specifically, the Composite column of the CSV dataset should have been loaded as float values instead of strings. Doing a check on Composite score column, reveals an error in the composite score of the state of Wyoming.

We have to fix this error before we can validate the CSV dataset against its source.

In [151]:
# Validate ACT CSV data with data from source @ https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows

# Set index (State) and rename columns of ACT CSV and source dataframes to facilitate validation later.

act_2017_columns = ['Participation', 'English', 'Math', 'Reading', 'Science', 'Composite']

df_act_2017.set_index('State', inplace=True)
df_act_2017.columns = act_2017_columns


def is_float(value):
    try:
        float(value)
        return True
    except ValueError:
        return False

def highlight_NaN_value(column):
    
    formats = ['' for i in range(column.size)]
    
    if column.name == 'Composite':
        return ['background-color: yellow' if not is_float(value) else '' for value in column]
    
    return formats

df_act_2017['is_composite_float'] = df_act_2017.apply(lambda row: is_float(row['Composite']), axis=1)
df_act_2017.tail().style.apply(highlight_NaN_value)

To fix this error, we change the Composite score of Wyoming to a numeric string (20.2) and convert the entire Composite score columnn to float values.


In [152]:
df_act_2017.loc['Wyoming','Composite'] = 20.2

df_act_2017['Composite'] = df_act_2017.apply(lambda row: float(row['Composite']), axis=1)

df_act_2017.drop('is_composite_float', axis=1, inplace=True)

df_act_2017.info()

<class 'pandas.core.frame.DataFrame'>
Index: 52 entries, National to Wyoming
Data columns (total 6 columns):
Participation    52 non-null object
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null float64
dtypes: float64(5), object(1)
memory usage: 3.9+ KB


Now we can validate the ACT CSV dataset against its source, and address any errors found.


In [153]:
# Load source SAT data from tab delimited file into dataframe.
df_act_2017_source = pd.read_csv('../data/act_2017_source.txt', sep='\t')
df_act_2017_source.set_index('State', inplace=True)
df_act_2017_source.columns = act_2017_columns
df_act_2017_source.rename(columns=lambda col_name: col_name + '_src', inplace=True) 

# Join ACT CSV and source dataframes on row index 'State' 
df_act_2017_combined = df_act_2017.join(df_act_2017_source, 'State')

df_act_2017_mismatched = df_act_2017_combined.loc[get_mismatched_indexes(df_act_2017_combined, act_2017_columns)]

def highlight_act_mismatched(column):
    
    formats = ['' for i in range(column.size)]
    
    if column.name in act_2017_columns:
        
        is_mismatched = column != df_act_2017_mismatched[column.name + '_src']

        return ['background-color: yellow' if mismatched else '' for mismatched in is_mismatched]
    
    return formats

df_act_2017_mismatched.style.apply(highlight_act_mismatched)

Cross-referencing the ACT source, we found 2 errors:
1. The Science score of Maryland should have been 23.2
2. The national Participation rate differs from the source due to a leading whitespace. This error will be resolved when converting the Participation column from string to int values.

#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

For the SAT dataset, we need to correct the Math score of Maryland to 524.
For the ACT dataset, we need to correct the Science score of Maryland to 23.2.
For both datasets, we need to convert Particatpation columns from string to float values.


In [154]:
def highlight_math_column(column):
    
    if column.name == 'Math':
        return ['background-color: yellow']
    else:
        return ['']

df_sat_2017.loc['Maryland','Math'] = 524
df_sat_2017.loc[['Maryland']].style.apply(highlight_column)

In [155]:
def highlight_science_column(column):
    
    if column.name == 'Science':
        return ['background-color: yellow']
    else:
        return ['']

df_act_2017.loc['Maryland','Science'] = 23.2
df_act_2017.loc[['Maryland']].style.apply(highlight_column)

#### 5. What are your data types? 
Display the data types of each feature. 

In [116]:
df_sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51 entries, Alabama to Wyoming
Data columns (total 4 columns):
Participation    51 non-null object
EBRW             51 non-null int64
Math             51 non-null int64
Total            51 non-null int64
dtypes: int64(3), object(1)
memory usage: 3.1+ KB


In [117]:
df_act_2017.info()

<class 'pandas.core.frame.DataFrame'>
Index: 52 entries, National to Wyoming
Data columns (total 6 columns):
Participation    52 non-null object
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null float64
dtypes: float64(5), object(1)
memory usage: 3.9+ KB


What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

The Participation column (from both datasets) should be of integer data type to reflect percentages of state participation rates.

Note: The Composite score column from the ACT dataset was converted from string to float data type earlier to facilitate validation against its source.

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [156]:
# Remove 

def to_int_column(df, col_name):
    df[col_name] = df.apply(lambda row: int(row[col_name].strip()[:-1]), axis=1)

to_int_column(df_sat_2017, 'Participation')
to_int_column(df_act_2017, 'Participation')

- Fix any individual values preventing other columns from being the appropriate type.

- Finish your data modifications by making sure the columns are now typed appropriately.

In [2]:
#code

- Display the data types again to confirm they are correct.

In [157]:
df_sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51 entries, Alabama to Wyoming
Data columns (total 4 columns):
Participation    51 non-null int64
EBRW             51 non-null int64
Math             51 non-null int64
Total            51 non-null int64
dtypes: int64(4)
memory usage: 3.3+ KB


In [158]:
df_act_2017.info()

<class 'pandas.core.frame.DataFrame'>
Index: 52 entries, National to Wyoming
Data columns (total 6 columns):
Participation    52 non-null int64
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null float64
dtypes: float64(5), int64(1)
memory usage: 4.1+ KB


#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [159]:
# Rename index and column names to lowercase. Append pfrefix 'sat_' or 'act_' to the respective datasets.  
df_sat_2017.index.name = 'state'
df_sat_2017.columns = ['sat_participation', 'sat_EBRW', 'sat_math', 'sat_total']

df_act_2017.index.name = 'state'
df_act_2017.columns = ['act_participation', 'act_english', 'act_math', 'act_reading', 'act_science', 'act_composite']

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature|Type|Dataset|Description|
|---|---|---|---|
|sat_participation|int|SAT|Percentage of students in the state taking the SAT college admissions test.| 
|sat_EBRW|int|SAT|Evidence-Based Reading Writing score of the SAT test.| 
|sat_math|int|SAT|Math score of the SAT test.| 
|sat_total|int|SAT|Total score of the SAT test (EBRW + Math score).| 
|act_participation|int|ACT|Percentage of students in the state taking the ACT college admissions test.| 
|act_english|float|ACT|English score of the ACT test.| 
|act_math|float|ACT|Math score of the ACT test.| 
|act_reading|float|ACT|Reading score of the ACT test.| 
|act_science|float|ACT|Science score of the ACT test.| 
|act_composite|float|ACT|Composite score of the ACT test (Average of English, Math, Reading and Science scores).| 

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [162]:
df_act_2017

act_participation  act_english  act_math  act_reading  \
state                                                                         
National                             60         20.3      20.7         21.4   
Alabama                             100         18.9      18.4         19.7   
Alaska                               65         18.7      19.8         20.4   
Arizona                              62         18.6      19.8         20.1   
Arkansas                            100         18.9      19.0         19.7   
California                           31         22.5      22.7         23.1   
Colorado                            100         20.1      20.3         21.2   
Connecticut                          31         25.5      24.6         25.6   
Delaware                             18         24.1      23.4         24.8   
District of Columbia                 32         24.4      23.5         24.9   
Florida                              73         19.0      19.4         21.0   
Georgia                              55         21.0      20.9         22.0   
Hawaii                               90         17.8      19.2         19.2   
Idaho                                38         21.9      21.8         23.0   
Illinois                             93         21.0      21.2         21.6   
Indiana                              35         22.0      22.4         23.2   
Iowa                                 67         21.2      21.3         22.6   
Kansas                               73         21.1      21.3         22.3   
Kentucky                            100         19.6      19.4         20.5   
Louisiana                           100         19.4      18.8         19.8   
Maine                                 8         24.2      24.0         24.8   
Maryland                             28         23.3      23.1         24.2   
Massachusetts                        29         25.4      25.3         25.9   
Michigan                             29         24.1      23.7         24.5   
Minnesota                           100         20.4      21.5         21.8   
Mississippi                         100         18.2      18.1         18.8   
Missouri                            100         19.8      19.9         20.8   
Montana                             100         19.0      20.2         21.0   
Nebraska                             84         20.9      20.9         21.9   
Nevada                              100         16.3      18.0         18.1   
New Hampshire                        18         25.4      25.1         26.0   
New Jersey                           34         23.8      23.8         24.1   
New Mexico                           66         18.6      19.4         20.4   
New York                             31         23.8      24.0         24.6   
North Carolina                      100         17.8      19.3         19.6   
North Dakota                         98         19.0      20.4         20.5   
Ohio                                 75         21.2      21.6         22.5   
Oklahoma                            100         18.5      18.8         20.1   
Oregon                               40         21.2      21.5         22.4   
Pennsylvania                         23         23.4      23.4         24.2   
Rhode Island                         21         24.0      23.3         24.7   
South Carolina                      100         17.5      18.6         19.1   
South Dakota                         80         20.7      21.5         22.3   
Tennessee                           100         19.5      19.2         20.1   
Texas                                45         19.5      20.7         21.1   
Utah                                100         19.5      19.9         20.8   
Vermont                              29         23.3      23.1         24.4   
Virginia                             29         23.5      23.3         24.6   
Washington                           29         20.9      21.9         2

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [ ]:
#Code:

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [7]:
#code

## 2018 Data Import and Cleaning

Links to the 2018 ACT and SAT data are provided in the README. These data live in PDFs, and so you'll get to enjoy practicing some *manual* data collection. Save these data as a CSV in your `data` directory, and import, explore, and clean these data in the same way you did above. **Make sure you comment on your steps so it is clear *why* you are doing each process**.

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [ ]:
#Code:

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [4]:
#code

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [ ]:
#Code:

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

Answer

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [2]:
#code

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2)) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        # Set titles, labels, etc here for each subplot

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

In [ ]:
# Code

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
# Code

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
# Code

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer:

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer:

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer:

##### Is it appropriate to compare *these* specific SAT and ACT math scores? 

Why or why not?

Answer:

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code:

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?